indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TasksAndWorkersFilteredByIp.java

```diff
@@ -0,0 +1,130 @@
+import io.druid.indexing.overlord.ImmutableWorkerInfo;
+import io.druid.indexing.overlord.TaskRunner;
+import io.druid.indexing.overlord.TaskRunnerWorkItem;
+import io.druid.indexing.overlord.TasksAndWorkers;
+import io.druid.indexing.overlord.WorkerTaskRunner;
+import io.druid.indexing.overlord.config.WorkerTaskRunnerConfig;
+import io.druid.indexing.worker.Worker;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+public class TasksAndWorkersFilteredByIp implements TasksAndWorkers
+{
+  private final WorkerTaskRunner delegate;
+  private final String ipPrefix;
+  private final Predicate<Task> taskPredicate;
+
+  public TasksAndWorkersFilteredByIp(WorkerTaskRunner delegate, String ipPrefix, Predicate<Task> taskPredicate)
+  {
+    this.delegate = delegate;
+    this.ipPrefix = ipPrefix;
+    this.taskPredicate = taskPredicate;
+  }
+
+  private Collection<Worker> filter(Collection<Worker> workers)
+  {
+    Collection<Worker> filtered = new ArrayList<>(workers.size());
+    for (Worker worker : workers) {
+      if (worker.getIp().startsWith(ipPrefix)) {
+        filtered.add(worker);
+      }
+    }
+    return filtered;
+  }
+
+  @Override
+  public Collection<ImmutableWorkerInfo> getWorkers()
+  {
+    Collection<ImmutableWorkerInfo> workers = delegate.getWorkers();
+    Collection<ImmutableWorkerInfo> filtered = new ArrayList<>(workers.size());
+    for (ImmutableWorkerInfo workerInfo : workers) {
+      if (workerInfo.getWorker().getIp().startsWith(ipPrefix)) {
+        filtered.add(workerInfo);
+      }
+    }
+    return filtered;
+  }
+
+  @Override
+  public Collection<Worker> getLazyWorkers()
+  {
+    return filter(delegate.getLazyWorkers());
+  }
+
+  @Override
+  public Collection<Worker> markWorkersLazy(final Predicate<ImmutableWorkerInfo> isLazyWorker, int maxWorkers)
+  {
+    Predicate<ImmutableWorkerInfo> isLazyWorkerWithProperIp = new Predicate<ImmutableWorkerInfo>()
+    {
+      @Override
+      public boolean apply(@Nullable ImmutableWorkerInfo input)
+      {
+        return isLazyWorker.apply(input) && input.getWorker().getIp().startsWith(ipPrefix);
+      }
+    };
+    return filter(delegate.markWorkersLazy(isLazyWorkerWithProperIp, maxWorkers));
+  }
+
+  @Override
+  public WorkerTaskRunnerConfig getConfig()
+  {
+    return delegate.getConfig();
+  }
+
+  @Override
+  public Collection<Task> getPendingTaskPayloads()
+  {
+    ImmutableList.Builder<Task> filtered = ImmutableList.builder();
+    for (Task task : delegate.getPendingTaskPayloads()) {
+      if (taskPredicate.apply(task)) {
+        filtered.add(task);
+      }
+    }
+    return filtered.build();
+  }
+
+  @Override
+  public Collection<? extends TaskRunnerWorkItem> getPendingTasks()
+  {
+    Collection<? extends TaskRunnerWorkItem> pendingTasks = delegate.getPendingTasks();
+    Set<String> pendingTaskIds = new HashSet<>();
```

Comment:
This one could be done with a combination of `com.google.common.collect.Collections2#transform` and filter, but I'm not sure if it would be much cleaner.